balancer/balanceraggregator/balanceraggregator.go

```diff
@@ -0,0 +1,286 @@
+	defer func() {
+		ba.inhibitChildUpdates.Store(false)
+		ba.updateState()
+	}()
+	var ret error
+	// Update/Create new children.
+	for _, endpoint := range state.ResolverState.Endpoints {
+		if _, ok := endpointSet.Get(endpoint); ok {
+			// Endpoint child was already created, continue to avoid duplicate
+			// update.
+			continue
+		}
+		endpointSet.Set(endpoint, nil)
+		var bal *balancerWrapper
+		if child, ok := ba.children.Get(endpoint); ok {
+			bal = child.(*balancerWrapper)
+		} else {
+			bal = &balancerWrapper{
+				childState: ChildState{Endpoint: endpoint},
+				ClientConn: ba.cc,
+				ba:         ba,
+			}
+			bal.Balancer = gracefulswitch.NewBalancer(bal, ba.bOpts)
+			ba.children.Set(endpoint, bal)
+		}
+		if err := bal.UpdateClientConnState(balancer.ClientConnState{
+			BalancerConfig: state.BalancerConfig,
+			ResolverState: resolver.State{
+				Endpoints:  []resolver.Endpoint{endpoint},
+				Attributes: state.ResolverState.Attributes,
+			},
+		}); err != nil && ret == nil {
+			// Return first error found, and always commit full processing of
+			// updating children. If desired to process more specific errors
+			// across all endpoints, caller should make these specific
+			// validations, this is a current limitation for simplicities sake.
+			ret = err
+		}
+	}
+	// Delete old children that are no longer present.
+	for _, e := range ba.children.Keys() {
+		child, _ := ba.children.Get(e)
+		bal := child.(balancer.Balancer)
+		if _, ok := endpointSet.Get(e); !ok {
+			bal.Close()
+			ba.children.Delete(e)
+		}
+	}
+	return ret
+}
+
+// ResolverError forwards the resolver error to all of the balancerAggregator's
+// children and sends a single synchronous update of the childStates at the end
+// of the ResolverError operation.
+func (ba *balancerAggregator) ResolverError(err error) {
+	ba.inhibitChildUpdates.Store(true)
+	defer func() {
+		ba.inhibitChildUpdates.Store(false)
+		ba.updateState()
+	}()
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.ResolverError(err)
+	}
+}
+
+func (ba *balancerAggregator) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	// UpdateSubConnState is deprecated.
+}
+
+func (ba *balancerAggregator) Close() {
+	for _, child := range ba.children.Values() {
+		bal := child.(balancer.Balancer)
+		bal.Close()
+	}
+}
+
+// updateState updates this component's state. It sends the aggregated state,
+// and a picker with round robin behavior with all the child states present if
+// needed.
+func (ba *balancerAggregator) updateState() {
+	if ba.inhibitChildUpdates.Load() {
+		return
+	}
+	readyPickers := make([]balancer.Picker, 0)
+	connectingPickers := make([]balancer.Picker, 0)
+	idlePickers := make([]balancer.Picker, 0)
+	transientFailurePickers := make([]balancer.Picker, 0)
+
+	childStates := make([]ChildState, 0, ba.children.Len())
+	ba.mu.Lock()
+	defer ba.mu.Unlock()
+	for _, child := range ba.children.Values() {
```

Comment:
I think you still have races here.

You don't take this lock in UpdateClientConnState, which mutates `ba.children`, and now you're reading the same map here.

Let's have the mutex just protect `ba.children` and `ba.children.childState` and make sure we don't make any calls across levels with it held?

Reply:
Sorry for all these synchronization correctness issues. I just took a look at Outlier Detection, and I guard the core data structures on updates from parent, updates from child, and the interval timer algorithm. I was initially worried about littering all the operations with mutexes but thinking about it these operations on the balancer type aren't on the RPC fast path, and don't happen at too high a frequency anyway where I think lock contention would be a problem.

Reply:
Let me think about weird behavioral splicing things that can happen here if I guard the children map in UpdateClientConnState.

Reply:
Fixed race as discussed offline.